jax/_src/random.py

```diff
@@ -935,30 +936,44 @@ def bernoulli(key: ArrayLike,
     shape: optional, a tuple of nonnegative integers representing the result
       shape. Must be broadcast-compatible with ``p.shape``. The default (None)
       produces a result shape equal to ``p.shape``.
+    mode: optional, "high" or "low" for how many bits to use when sampling.
+      default='low'.
 
   Returns:
     A random array with boolean dtype and shape given by ``shape`` if ``shape``
     is not None, or else ``p.shape``.
   """
   if shape is not None:
     shape = core.canonicalize_shape(shape)
+  if mode not in ['high', 'low']:
+    raise ValueError(f"got {mode=}, expected 'high' or 'low'")
   key, _ = _check_prng_key("bernoulli", key)
   dtype = dtypes.canonicalize_dtype(lax.dtype(p))
   if not jnp.issubdtype(dtype, np.floating):
     msg = "bernoulli probability `p` must have a floating dtype, got {}."
     raise TypeError(msg.format(dtype))
   p = lax.convert_element_type(p, dtype)
-  return _bernoulli(key, p, shape)
+  return _bernoulli(key, p, shape, mode=mode)
 
-@partial(jit, static_argnums=(2,))
-def _bernoulli(key, p, shape) -> Array:
+
+@partial(jit, static_argnames=['shape', 'mode'])
+def _bernoulli(key: Array, p: Array, shape: Shape | None, mode: str) -> Array:
   if shape is None:
     # TODO: Use the named part of `p` as well
     shape = np.shape(p)
   else:
     _check_shape("bernoulli", shape, np.shape(p))
 
-  return uniform(key, shape, lax.dtype(p)) < p
+  if mode == 'high':
+    key1, key2 = split(key)
+    dtype = lax.dtype(p)
+    # spacing between uniform values is 2 ** -n_mantissa
+    delta = 2 ** -dtypes.finfo(dtype).nmant
+    u1 = uniform(key1, shape, lax.dtype(p))
```

Comment:
I think you can avoid the extra split step by doing:
```
u1, u2 = uniform(key, (2,) + shape, lax.dtype(p))
return u2 * delta < p - u1
```

Reply:
Is slicing really better than splitting here?

Reply:
I think yes, consider that splitting then uniform is two sequential operations whereas slicing can be fully fused into a single operation.

Reply:
Done